### PR TITLE
Chat widget changes to show the queue part

### PR DESCRIPTION
**Pull request by Kusum-Dhoundiyal**
### What do you want to achieve?
-Chat widget changes to show the queue part

### PR Checklist

Screenshot 2020-09-15 at 7 48 35 PM
- [x] I have tested it locally and all functionalities are working fine.
- [ ] I have compared it with mocks and all design elements are the same.
- [ ] I have tested it in IE Browser.

### How was the code tested?

-

### What new thing you came across while writing this code? 
-

### In case you fixed a bug then please describe the root cause of it? 
-

NOTE: Make sure you're comparing your branch with the correct base branch

### CODE REVIEW

File: webplugin/template/mck-sidebox.html [hunk 1]
```diff
@@ -304,7 +304,7 @@ <h3>No more messages!</h3>
 				</div>
 
 				<div class="mck-conversation-comment n-vis" id="csat-3"></div>
-
+                <div id = "mck-waiting-queue">You are currently in the waiting queue, our agents will get back you shortly. </div>
```
Comment (devashishmamgain) [inline webplugin/template/mck-sidebox.html:307]: please make a habit of using MCK_LABELS

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/km-utils.js [hunk 2]
```diff
@@ -57,6 +57,18 @@ KommunicateConstants = {
     CONVERSATION_CLOSED_STATUS : 'closed',
     CONVERSATION_RESOLVED_STATUS : 'Resolved',
     CONVERSATION_OPEN_STATUS : 'Open',
+    CONVERSATION_WAITING_STATUS:'WAITING',
+    CONVERSATION_STATE: {
```
Comment (devashishmamgain) [inline webplugin/js/app/km-utils.js:61]: these constants are present in km-conversation-helper.js

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/css/app/mck-sidebox-1.0.css [hunk 3]
```diff
@@ -5404,7 +5404,15 @@ div[name="message"]:last-child {
 	color: #676262;
 	font-size: 14px;
 }
-
+div#mck-waiting-queue{
+	color: #5C6677;
```
Comment (akshatdhabi) [inline webplugin/css/app/mck-sidebox-1.0.css:5408]: Check code formatting

**Comment by akshatdhabi** [comment]
image

image

image



Please check the dropdown UI and also default is supposed to be set to 5

File: webplugin/js/app/km-utils.js [hunk 4]
```diff
@@ -57,6 +57,7 @@ KommunicateConstants = {
     CONVERSATION_CLOSED_STATUS : 'closed',
     CONVERSATION_RESOLVED_STATUS : 'Resolved',
     CONVERSATION_OPEN_STATUS : 'Open',
+    CONVERSATION_WAITING_STATUS:'WAITING',
```
Comment (devashishmamgain) [inline webplugin/js/app/km-utils.js:60]: others are in lowercase, this one is in CAPS, whats the reason?

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 5]
```diff
@@ -9806,6 +9808,7 @@ var KM_ATTACHMENT_V2_SUPPORTED_MIME_TYPES = ["application","text","image"];
                                 MCK_GROUP_MAP && MCK_GROUP_MAP[resp.message.groupId] && (MCK_GROUP_MAP[resp.message.groupId].metadata.CONVERSATION_STATUS = Kommunicate.conversationHelper.status.OPEN);
                             }
                             KommunicateUI.handleConversationBanner();
+                            resp.message.metadata.KM_STATUS == KommunicateConstants.CONVERSATION_WAITING_STATUS   ?KommunicateUI.handleWaitingQueueMessage(true):KommunicateUI.handleWaitingQueueMessage(false);
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js:9811]: simplify this

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate-client.js [hunk 6]
```diff
@@ -94,6 +94,7 @@ Kommunicate.client={
                     if (typeof callback == 'function') {
                         callback(response.data.value);
                     }
+                    response && response.data.metadata.CONVERSATION_STATUS ==Kommunicate.conversationHelper.status.WAITING ?KommunicateUI.handleWaitingQueueMessage(true):KommunicateUI.handleWaitingQueueMessage(false);
```
Comment (ab14bhardwaj) [inline webplugin/js/app/kommunicate-client.js:97]: let waitingStatus = response && response.data.metadata.CONVERSATION_STATUS ==Kommunicate.conversationHelper.status.WAITING;
KommunicateUI.handleWaitingQueueMessage(waitingStatus);

File: webplugin/js/app/kommunicate-client.js [hunk 7]
```diff
@@ -94,6 +94,8 @@ Kommunicate.client={
                     if (typeof callback == 'function') {
                         callback(response.data.value);
                     }
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate-client.js:96]: Same check is present at 3 places, see if directly the status value can be passed to the function. And that function can internally check this part

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

**Review by devashishmamgain — APPROVED** [review]

**Review by ab14bhardwaj — DISMISSED** [review]

**Review by devashishmamgain — APPROVED** [review]

File: webplugin/js/app/kommunicate-ui.js [hunk 8]
```diff
@@ -588,7 +588,34 @@ handleAttachmentIconVisibility : function(enableAttachment, msg, groupReloaded)
             startConversationButton.classList.add('force-n-vis');
             hasMultipleConversations ? backButton.classList.remove('force-n-vis') : backButton.classList.add('force-n-vis')
         }
-    }
+    },
+        handleWaitingQueueMessage: function (data) {
+            let groupId = data && data.clientGroupId;
+            let waitingStatus = data && data.metadata.CONVERSATION_STATUS == Kommunicate.conversationHelper.status.WAITING;
+            mckUtils.ajax({
```
Comment (ab14bhardwaj) [inline webplugin/js/app/kommunicate-ui.js:595]: @Kusum-Dhoundiyal We use below function to make requests in web-SDK.
``` 
 window.Applozic.ALApiService.ajax({
```

File: webplugin/js/app/kommunicate-ui.js [hunk 9]
```diff
@@ -588,7 +588,34 @@ handleAttachmentIconVisibility : function(enableAttachment, msg, groupReloaded)
...
+                    if (waitingStatus) {
+                        kommunicateCommons.modifyClassList({
+                            id: ["mck-waiting-queue"]
+                        }, "vis", "n-vis");
+                    } else {
+                        kommunicateCommons.modifyClassList({
+                            id: ["mck-waiting-queue"]
+                        }, "n-vis", "vis");
+                    }
+
```
Comment (ab14bhardwaj) [inline webplugin/js/app/kommunicate-ui.js:612]: @Kusum-Dhoundiyal Check the response code is "success" or not if coming.
If nothing is coming to verify if the API was successful or not then you can resolve my comment.

File: webplugin/js/app/kommunicate-ui.js [hunk 10]
```diff
@@ -588,7 +588,34 @@ handleAttachmentIconVisibility : function(enableAttachment, msg, groupReloaded)
...
+                },
+                error: function (err) {
+                    console.log('Error while fetching waiting list', err);
```
Comment (ab14bhardwaj) [inline webplugin/js/app/kommunicate-ui.js:615]: We remove console.log while minification, so please throw the error.

File: webplugin/template/mck-sidebox.html [hunk 11]
```diff
@@ -304,7 +304,9 @@ <h3>No more messages!</h3>
 				</div>
 
 				<div class="mck-conversation-comment n-vis" id="csat-3"></div>
-
+				<div id = "mck-waiting-queue"  class="n-vis">
+					<span id="mck-waiting-queue-1"></span> <span id="mck-waiting-queue-2" ></span><span id="mck-waiting-queue-3" ></span>
+				</div>
```
Comment (ab14bhardwaj) [inline webplugin/template/mck-sidebox.html:309]: Please format the code, if it's formatted then resolve the comment.

Comment (ab14bhardwaj) [reply webplugin/template/mck-sidebox.html:309]: Plus give a proper name to the ids by the name you have given I am not able to understand anything.

File: webplugin/js/app/labels/default-labels.js [hunk 12]
```diff
@@ -161,5 +161,9 @@ Kommunicate.defaultLabels = {
 
         },
         'attachment': 'You have an attachment.'
-    }
+    },
+    'waiting.queue':{
+        'first.Part':'You are currently',
+        'waiting.queue.number':'5',
+        'last.part': ' in the waiting queue, our agents will get back you shortly.'},
```
Comment (ab14bhardwaj) [inline webplugin/js/app/labels/default-labels.js:168]: "waiting.queue.message" or "waiting.queue.heading"
The label name should be self-explanatory as the customers will use it.

NOTE: Please create the PR in docs repo also to add these labels.

File: webplugin/js/app/kommunicate-ui.js [hunk 13]
```diff
@@ -589,15 +589,33 @@ handleAttachmentIconVisibility : function(enableAttachment, msg, groupReloaded)
...
+            let groupId = data && data.clientGroupId;
+            let waitingStatus = data && data.metadata.CONVERSATION_STATUS == Kommunicate.conversationHelper.status.WAITING;
+            mckUtils.ajax({
+                type: 'GET',
+                url: MCK_BASE_URL + '/rest/ws/group/waiting/list',
+                global: false,
+                contentType: 'application/json',
+                success: function (res) {
+                    WAITING_QUEUE = res.response;
+                    document.getElementById('mck-waiting-queue-2') && (document.getElementById('mck-waiting-queue-2').innerHTML = "#" + parseInt(WAITING_QUEUE.indexOf(parseInt(groupId)) + 1));
```
Comment (ab14bhardwaj) [inline webplugin/js/app/kommunicate-ui.js:602]: Check if the status code or message "success" is available in the API to verify if results are properly fetched or not.

**Review by ab14bhardwaj — APPROVED** [review]

**Review by akshatdhabi — APPROVED** [review]

**Comment by ab14bhardwaj** [comment]
@devashishmamgain Merging the PR as you review it earlier, the review was dismissed automatically as per Kusum.

**Pull request merged**